criu/kerndat.c

```diff
@@ -531,6 +571,31 @@ static bool kerndat_has_memfd_hugetlb(void)
 	return 0;
 }
 
+static bool kerndat_has_memfd_secret(void)
+{
+#if defined(CONFIG_HAS_MEMFD_SECRET)
+	int ret;
+
+	ret = memfd_secret(0);
+
+	if (ret > 0) {
+		kdat.has_memfd_secret = true;
+		close(ret);
+	} else if (ret == -1 && (errno == ENOSYS || errno == EINVAL || errno == EMFILE || errno == ENOMEM)) {
+		kdat.has_memfd_secret = false;
+	} else {
+		pr_perror("Unexpected error from memfd_secret(0)");
+		return -1;
+	}
```

Comment:
I think it's better to structure the conditions here so that we return on errors and continue when everything is fine. And we can call `kerndat_get_secretmem_dev` from this function or probably even merge these two.
E.g something along these lines:

```
kdat.has_memfd_secret = false;
fd = memfd_secret(0);
if (fd < 0) {
	if (errno == ENOSYS)
		return 0;
	return fd;
}

ret = kerndat_get_secretmem_dev(fd);
if (ret) {
	close(fd);
	return ret;
}
...
```

Reply:
@rppt I have updated the [`kerndat_get_secretmem_dev()`](https://github.com/checkpoint-restore/criu/blob/2e8fba61b9683904a1faf45ac53fef38dfdb648b/criu/kerndat.c#L261) and [`kerndat_has_memfd_secret()`](https://github.com/checkpoint-restore/criu/blob/2e8fba61b9683904a1faf45ac53fef38dfdb648b/criu/kerndat.c#L576) according to the previously requested changes (now resolved) . And I'm not sure why we need this change? Is it because, we are calling `memfd_secret()` syscall twice in both functions?

Reply:
done.